ports/esp32/esp32_can.c

```diff
@@ -0,0 +1,759 @@
+    if (self->config->initialized) {
+        state = _esp32_hw_can_get_status().state;
+    }
+    return mp_obj_new_int(state);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp32_hw_can_state_obj, esp32_hw_can_state);
+
+// info() -- Get info about error states and TX/RX buffers
+STATIC mp_obj_t esp32_hw_can_info(size_t n_args, const mp_obj_t *args) {
+    esp32_can_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_obj_list_t *list;
+    if (n_args == 1) {
+        list = MP_OBJ_TO_PTR(mp_obj_new_list(7, NULL));
+    } else {
+        if (!mp_obj_is_type(args[1], &mp_type_list)) {
+            mp_raise_TypeError(NULL);
+        }
+        list = MP_OBJ_TO_PTR(args[1]);
+        if (list->len < 7) {
+            mp_raise_ValueError(NULL);
+        }
+    }
+    twai_status_info_t status = _esp32_hw_can_get_status();
+    list->items[0] = MP_OBJ_NEW_SMALL_INT(status.tx_error_counter);
+    list->items[1] = MP_OBJ_NEW_SMALL_INT(status.rx_error_counter);
+    list->items[2] = MP_OBJ_NEW_SMALL_INT(self->num_error_warning);
+    list->items[3] = MP_OBJ_NEW_SMALL_INT(self->num_error_passive);
+    list->items[4] = MP_OBJ_NEW_SMALL_INT(self->num_bus_off);
+    list->items[5] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_tx);
+    list->items[6] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_rx);
+    return MP_OBJ_FROM_PTR(list);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp32_hw_can_info_obj, 1, 2, esp32_hw_can_info);
+
+// any() - return `True` if any message waiting, else `False`
+STATIC mp_obj_t esp32_hw_can_any(mp_obj_t self_in) {
+    twai_status_info_t status = _esp32_hw_can_get_status();
+    return mp_obj_new_bool((status.msgs_to_rx) > 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp32_hw_can_any_obj, esp32_hw_can_any);
+
+// send([data], id, *, timeout=0, rtr=false, extframe=false)
+STATIC mp_obj_t esp32_hw_can_send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_data, ARG_id, ARG_timeout, ARG_rtr, ARG_extframe };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_data,     MP_ARG_REQUIRED | MP_ARG_OBJ,   {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_id,       MP_ARG_REQUIRED | MP_ARG_INT,   {.u_int = 0} },
+        { MP_QSTR_timeout,  MP_ARG_KW_ONLY | MP_ARG_INT,    {.u_int = 0} },
+        { MP_QSTR_rtr,      MP_ARG_KW_ONLY | MP_ARG_BOOL,   {.u_bool = false} },
+        { MP_QSTR_extframe, MP_ARG_BOOL,                    {.u_bool = false} },
+    };
+
+    // parse args
+    esp32_can_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // populate message
+    twai_message_t tx_msg;
+
+    size_t length;
+    mp_obj_t *items;
+    mp_obj_get_array(args[ARG_data].u_obj, &length, &items);
+    if (length > CAN_MAX_DATA_FRAME) {
+        mp_raise_ValueError("CAN data field too long");
+    }
+    tx_msg.data_length_code = length;
+    tx_msg.identifier = args[ARG_id].u_int;
```

Comment:
checking for proper frame ID. You are forcefully modifying the frame id that a user is passing instead of kicking back an error telling them that the frame id is out of bounds. you are also setting `tx_msg.identifier`  before  a check is done as to the id type being used. Seems rather odd to do that twice.

One of the things that is done in DBC files is the setting of bit 30 in the frame id to identify the frame id bit depth. If it is a 1 it's extended and if it is a 0 then it is not. might want to consider using that. just an idea.

Reply:
>    tx_msg.identifier = args[ARG_id].u_int;

Oh, that's literally left by chance, I'll fix it.

As for rising exception -- I agree with it, but stm32 implementation [has literally the same code](https://github.com/micropython/micropython/blob/b8982ec5f90a5aa9d802e9d54912dc4e7ba0916a/ports/stm32/pyb_can.c#L536), that only converts identifiers to correct intervals (30 or 11 bits). It is even [tested in the same way](https://github.com/micropython/micropython/blob/b8982ec5f90a5aa9d802e9d54912dc4e7ba0916a/tests/pyb/can.py#L49)

Reply:
I am in agreement with you o a lot of this because tying to match the STM32 API is crazy because a lot is going to get lost.

For example if one was to keep the API the same then the loopback mode would need to be removed from the STM32 CAN. There is no loopback mode on the ESP32. nor is there a silent loopback mode.


TWAI_MODE_NORMAL
TWAI_MODE_NO_ACK
TWAI_MODE_LISTEN_ONLY

Those are the modes that are available for the ESP32

Alerts and states. ESP32 has these ones that are going to end up going unused. 

CAN_STATE_RECOVERING

CAN_ALERT_BUS_OFF
CAN_ALERT_TX_FAILED
CAN_ALERT_BUS_ERROR
CAN_ALERT_ARB_LOST
CAN_ALERT_BUS_RECOVERED
CAN_ALERT_RECOVERY_IN_PROGRESS
CAN_ALERT_TX_SUCCESS
CAN_ALERT_TX_IDLE
TWAI_ALERT_PERIPH_RESET

I believe the ESP32 also has a pin that gets set to either a high or a low I cannot remember which and that pin can get connected to an input pin that can wake the ESP32. That pin can be read to know when the ESP32 should go to sleep if the bus is not active. IDK if that feature is available or not. That right there is a damned handy thing to have if say you are developing something for a car.

I don't see how they can be made to have a  common API.